<a href="https://colab.research.google.com/github/Nikkuchan/CpE-AIML/blob/main/58089_PrelimPS_Padilla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Topic02a : Prelim Problem Set I

Padilla, Jasmine Clare B.

## Case 1
Represent the following representations into its vectorized form using LaTeX.
> **Problem 1.a. System of Linear Equations**
$$
\left\{
    \begin{array}\\
        -y+z=\frac{1}{32}\\ 
        \frac{1}{2}x -2y=0 \\
        -x + \frac{3}{7}z=\frac{4}{5}
    \end{array}
\right. $$
> **Problem 1.b. Linear Combination**
$$  \cos{(\theta)}\hat{i} + \sin{(\theta)}\hat{j} - \csc{(2\theta)}\hat{k}$$
> **Problem 1.c. Scenario**
>
>A conference has 200 student attendees, 45 professionals, and has 15 members of the panel. There is a team of 40 people on the organizing committee. Represent the *percent* composition of each *attendee* type of the conference in matrix form.

Express your answers in LaTeX in the answer area.


Problem 1.a

$$\begin{bmatrix}0&-1&1\\\frac{1}{2}&-2&0\\-1&0&\frac{3}{7}\end{bmatrix}=\begin{bmatrix}\frac{1}{32}\\0\\\frac{4}{5}\end{bmatrix}$$

Problem 1.b

$$\begin{bmatrix}\cos{(\theta)}i & \sin{(\theta)}j & - \csc{(2\theta)}k\end{bmatrix}$$


Problem 1.c

$$Conference = \left[\begin{matrix}\frac{2}{3} & \frac{3}{20} & \frac{1}{20} & \frac{2}{15}\end{matrix}\;\middle|\;\begin{matrix}1 \end{matrix}\right]$$

# Case 2
> **Problem 2.a: Vector Magnitude**

>The magnitude of a vector is usually computed as:
$$||v|| = \sqrt{a_0^2 + a_1^2 + ... +a_n^2}$$
Whereas $v$ is any vector and $a_k$ are its elements wherein $k$ is the size of $v$.
Re-formulate $||v||$ as a function of an inner product. Further discuss this concept and provide your user-defined function.

> **Problem 2.b: Angle Between Vectors**

> Inner products can also be related to the Law of Cosines. The property suggests that:
$$u\cdot v = ||u||\cdot||v||\cos(\theta)$$
Whereas $u$ and $v$ are vectors that have the same sizes and $\theta$ is the angle between $u$ and $v$.

> Explain the behavior of the dot product when the two vectors are perpendicular and when they are parallel.


##Problem 2.a: Vector Magnitude

The magnitude of a vector is determined by its length. The length of a vector must be calculated first before computing its magnitude. Velocity, displacement, force, momentum, and other quantities with direction are examples of vector quantities.

In [ ]:
import numpy as np

def vect_mag(k):
  return np.sqrt(sum(a**2 for a in k))

vector = np.array([1, 2, 3, 4, 5])
v = vect_mag(vector)

print(vector, v, sep = '\n')

[1 2 3 4 5]
7.416198487095663


## Problem 2.b: Angle Between Vectors

Given that vectors are quantities with direction, there would be something that would indicate that direction. In this case, it's an angle. 

The dot product when two vectors are perpendicular is null as cos(90) = 0.
$$u\cdot v = ||u||\cdot||v||\cos(90)$$
$$u\cdot v = ||u||\cdot||v||0$$
$$u\cdot v = 0$$

The dot product when two vectors are parallel is just the dot product of two vectors as cos(0) = 1.
$$u\cdot v = ||u||\cdot||v||\cos(0)$$
$$u\cdot v = ||u||\cdot||v||1$$
$$u\cdot v = u\cdot v$$


In [ ]:
def angle_vectors(u, v):
  inner = np.inner(u, v)
  norms = np.linalg.norm(u) * np.linalg.norm(v)
  cos = inner / norms
  rad = np.arccos(np.clip(cos, -1.0, 1.0))
  deg = np.rad2deg(rad)
  print("Radiant: ", rad)
  print("Degree: ", deg)
  return deg, rad

u = np.array([1, 2])
v = np.array([-1, -2])
print(u,v)

angle_vectors(u,v)

[1 2] [-1 -2]
Radiant:  3.1415926325163688
Degree:  179.99999879258172


(179.99999879258172, 3.1415926325163688)

# Case 3
For the final cases analysis we will be looking at series of equations building up a single feed-forward computation of a logistic regression. The case will not require you to learn fully what is logistic regression. 

$$X = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix} \text{, } 
Y = \begin{bmatrix} 
y^{(1)} \\ 
y^{(2)} \\
\vdots \\
y^{(m)} \\
\end{bmatrix} \text{, and } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} $$
The dataset $X$ has $m$ entries with $n$ features while $Y$ is the vector containing the groud truths of a the entries of $X$, and $\theta$ are the parameters or weights of the vectors. We first compute the vector product of the dataset and the parameters as:
$$ z = x^{(i)}\theta^{(i)} = X\cdot \theta\\_{\text{Eq. 3.1}}$$
We then solve for the hypothesis of the logistic regression alogrithm as:

$$ h_\theta(x) = g(z)\\_{\text{Eq. 3.2}}$$

Where $g$ is an acitvation function that maps the values of the hypothesis vector between a range of 0 and 1. We computed the activation as a sigmoid function:
$$g(z) = \frac{1}{1+e^{-z}}\\_{\text{Eq. 3.3}}$$
Finally we compute the loss of the logistic regression algorithm using $J$. Wheras $J(\theta)$ is a function that computes the logistic loss of the hypothesis with respect to the ground truths $y$. it is then computed as:
$$J(\theta) = \frac{1}{m} \sum^m_{i=0}=[-y^{(i)}\log({h_{\theta}(x^{(i)})})-(1-y^{(i)})\log(1-h_{\theta}(x^{(i)}))]\\_{\text{Eq. 3.4}}$$

> **Problem 3.a: Matrix Equivalences**

> In Eq. 1, $z$ can also be solved as $X \cdot \theta$ which is the vectorized form of $x^{(i)}\theta^{(i)}$. However, it can also be expressed as $\theta^T\cdot X$. Prove the equality of $X \cdot \theta$ with $\theta^T\cdot X$ in this case.

> **Problem 3.b: Matrix Shapes**

> Determine the shape of $h_\theta$ if $X$ has a shape of $(300,5)$.

> **Problem 3.c: Vectorization**

> Express $J(\theta)$ into its vectorized form.

> **Problem 4.c: Computational Programming (Also Laboratory 2)**

> Encode Equations 3.1 to 3.4 as the class `LRegression` wherein:

> * `LRegression` should be instantiated with a dataset $X$, a ground truth vector $y$, and a parameter vector $\theta$. Each parameter should have a data type of `numpy.array`.
> * It should further have `methods`reflecting to at least the four (4) aforementioned equations. Each should have a return value.


##Problem 3.a: Matrix Equivalence

Unlike scalar multiplication, multiplication involving matrices is not commutative. However, commutative property is allowed when getting the dot product between two vectors.


It means that $A\cdot B = B \cdot A $ can be applied even if one of the vectors is transposed. 

In problem 3.a, it is required to prove the equality of $X \cdot \theta$ with $\theta^T\cdot X$ given that $z$ can also be solved as $X \cdot \theta$ in equation 3.1
$$X = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix} \text{, } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} $$

By computing the vector product of the dataset and the parameters $$ z = X\cdot \theta$$ we get:

$$z = \begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta-\\
\vdots \\
— (x^{(m)})^T\theta— \\
\end{bmatrix}$$


By applying commutative property to $ z = X\cdot \theta$ which is the vectorized form of $x^{(i)^T} \theta^{(i)}$ , we can say that it is equal to 

$$ z = \theta^T\cdot X$$ 

$$z = \begin{bmatrix} 
—\theta^T(x^{(1)})— \\ 
—\theta^T(x^{(2)})-\\
\vdots \\
—\theta^T(x^{(m)})— \\
\end{bmatrix}$$

Hence, 
$$z = \begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta-\\
\vdots \\
— (x^{(m)})^T\theta— \\
\end{bmatrix} 
= \begin{bmatrix} 
—\theta^T(x^{(1)})— \\ 
—\theta^T(x^{(2)})-\\
\vdots \\
—\theta^T(x^{(m)})— \\
\end{bmatrix} $$

## Problem 3.b: Matrix Shapes

## Problem 3.c: Vectorization

$$\begin{bmatrix} 
\frac{\partial{J}}{\partial{\theta}_0} \\ 
\frac{\partial{J}}{\partial{\theta}_1} \\
\frac{\partial{J}}{\partial{\theta}_2} \\
\vdots \\
\frac{\partial{J}}{\partial{\theta}_n} \\
\end{bmatrix} = 
\frac{1}{m}
\begin{bmatrix} 
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\vdots \\
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\end{bmatrix} 
$$

$$
\frac{\partial{J}(\theta)}{\partial{\theta}_0} =
\frac{1}{m}
\sum\limits_{i=1}^m (h_0(x^{(i)})-y^{(i)})x^{(i)}_j 
$$
$$
\frac{\partial{J}(\theta)}{\partial{\theta}_0} = \left (
\frac{1}{m}
\sum\limits_{i=1}^m (h_0(x^{(i)})-y^{(i)})x^{(i)}_j \right) +
\frac{\lambda}{m} \theta_j
$$